src/appAccess.ts

```diff
@@ -0,0 +1,288 @@
+          : (references[lAppIndex - 1].appIndex as bigint);
+      references.push({ locals: { address: lAddress, appIndex: app } });
+    }
+    const box = item.get('b') as Map<string, unknown> | undefined;
+    if (box) {
+      const bAppIndex = ensureSafeUnsignedInteger(box.get('i') ?? 0);
+      const name = box.get('n') as Uint8Array;
+      if (!name) {
+        throw new Error(`Box missing name: ${box}`);
+      }
+      const app =
+        bAppIndex === 0
+          ? BigInt(0)
+          : (references[bAppIndex - 1].appIndex as number | bigint);
+      references.push({ box: { appIndex: app, name } });
+    }
+  }
+  return references;
+}
+
+/**
+ * foreignArraysToResourceReferences makes a single array of ResourceReferences from various foreign resource arrays.
+ * Note, runtime representation of ResourceReference uses addresses, app and asset identifiers, not indexes.
+ *
+ * @param accounts - optional array of accounts
+ * @param foreignAssets - optional array of foreign assets
+ * @param foreignApps - optional array of foreign apps
+ * @param holdings - optional array of holdings
+ * @param locals - optional array of locals
+ * @param boxes - optional array of boxes
+ */
+export function foreignArraysToResourceReferences({
+  appIndex,
+  accounts,
+  foreignAssets,
+  foreignApps,
+  holdings,
+  locals,
+  boxes,
+}: {
+  appIndex: bigint | number;
+  accounts?: ReadonlyArray<string | Address>;
+  foreignAssets?: ReadonlyArray<number | bigint>;
+  foreignApps?: ReadonlyArray<number | bigint>;
+  holdings?: ReadonlyArray<HoldingReference>;
+  locals?: ReadonlyArray<LocalsReference>;
+  boxes?: ReadonlyArray<BoxReference>;
+}): Array<ResourceReference> {
+  const accessList: Array<ResourceReference> = [];
+  function ensureAddress(addr: string | Address) {
+    let addr2: Address;
+    if (typeof addr === 'string') {
+      if (addr === '') {
+        return;
+      }
+      addr2 = Address.fromString(addr);
+    } else {
+      addr2 = addr;
+    }
+    if (addr2.equals(Address.zeroAddress())) {
+      return;
+    }
+    let addrFound = false;
+    for (const rr of accessList) {
+      if (!rr.address) {
+        continue;
+      }
+      let rrAddress = rr.address;
+      if (typeof rr.address === 'string') {
+        rrAddress = Address.fromString(rr.address);
+      }
+      if ((rrAddress as Address).equals(addr2)) {
+        addrFound = true;
+        break;
+      }
+    }
+    if (!addrFound) {
+      accessList.push({ address: addr });
+    }
+  }
+  function ensureAsset(asset: number | bigint) {
+    let assetFound = false;
+    for (const rr of accessList) {
+      if (rr.assetIndex && rr.assetIndex === asset) {
+        assetFound = true;
+        break;
+      }
+    }
+    if (!assetFound) {
+      accessList.push({ assetIndex: asset });
+    }
+  }
+  function ensureApp(app: number | bigint) {
+    let appFound = false;
+    for (const rr of accessList) {
+      if (rr.appIndex && rr.appIndex === app) {
```

Comment:
```suggestion
      if (rr.appIndex === app) {
```

Reply:
disagree, we do not want zero appIndex to appear in the list

Reply:
This would only matter if `app` is 0, which means someone asked to convert a foreign array that contains a 0.  With this check in place `appFound` is going be false, and you'll end up pushing a `{appIndex: 0}` on line 263.

Your comment makes it sound like what you _don't_ want that though.

Reply:
right, I misread